packages/build/src/lib/schema-parser/middleware/checkCache.ts

```diff
@@ -0,0 +1,169 @@
+  timeInterval?: string;
+  cacheTableName?: string;
+  urlPath?: string;
+}
+export class CheckCache extends SchemaParserMiddleware {
+  public async handle(schemas: RawAPISchema, next: () => Promise<void>) {
+    // check .yaml file has cache configuration
+    const caches = schemas?.cache;
+    const metadata = schemas.metadata;
+    const isUsedCache = metadata?.[CACHE_METADATA_NAME]?.['isUsedTag'];
+    if (!isUsedCache && !caches) return next();
+
+    // throw if cache not used in .sql file
+    if (isUsedCache && !caches) {
+      throw new ConfigurationError(
+        `{% cache %} tag was used in SQL file, but the cache configurations was not found in schema "${schemas.urlPath}".`
+      );
+    }
+    if (!isUsedCache && caches) {
+      throw new ConfigurationError(
+        `Can not configure the cache setting in schema "${schemas.urlPath}", {% cache %} tag not been used in SQL file.`
+      );
+    }
+    this.checkCacheTableName(schemas);
+    this.checkSQLHasValue(schemas);
+    this.assignCacheProfile(schemas);
+    this.checkRefreshSettings(schemas);
+    this.checkIndexesHasValue(schemas);
+
+    await next();
+  }
+
+  private checkRefreshSettings(schemas: RawAPISchema) {
+    const { cache: caches, urlPath } = schemas;
+    caches?.forEach((cache): void => {
+      // refreshTime and refreshExpression can not be set at the same time
+      const { refreshTime, refreshExpression, cacheTableName } = cache;
+      if (refreshTime && refreshExpression) {
+        throw new ConfigurationError(
+          `Can not set "refreshTime" and "refreshExpression" at the same time in cache "${cache.cacheTableName}" of schema "${schemas.urlPath}"`
+        );
+      }
+      // check "every" of "refreshTime" and "refreshExpression" are valid
+      if (refreshTime) {
+        const refreshOption: RefreshOption = {
+          intervalType: 'refreshTime',
+          timeInterval: refreshTime.every,
+          cacheTableName,
+          urlPath,
+        };
+        this.checkRefreshInterval(refreshOption);
+      }
+      if (refreshExpression) {
+        const refreshOption: RefreshOption = {
+          intervalType: 'refreshExpression',
+          timeInterval: refreshExpression.every,
+          cacheTableName,
+          urlPath,
+        };
+        this.checkRefreshInterval(refreshOption);
+      }
+    });
+  }
+
+  private checkRefreshInterval(checkOption: RefreshOption) {
+    const { intervalType, timeInterval, cacheTableName, urlPath } = checkOption;
+    const interval = ms(timeInterval as StringValue);
+    if (isNaN(interval)) {
+      throw new ConfigurationError(
+        `The "every" of "${intervalType}" in cache "${cacheTableName}" of schema "${urlPath}" is invalid: Invalid time string to convert.`
+      );
+    }
+    if (interval < 0) {
+      throw new ConfigurationError(
+        `The "every" of "${intervalType}" in cache "${cacheTableName}" of schema "${urlPath}" is invalid: Time can not be negative.`
+      );
+    }
```

Comment:
Because these codes still have some bad smell for using function arguments, so after discussion with @fredalai & @JSYOU, we found a great way to omit `if (refreshTime)` and `if (refreshExpression)`, also including omit the two errors in `checkRefreshInterval`, and missed checking `every` is undefined or not to throw specific error message 


```ts
  for (const cache of caches!) {
      // refreshTime and refreshExpression can not be set at the same time
      const { refreshTime, refreshExpression, cacheTableName } = cache;
      if (!refreshTime && !refreshExpression) return;

      if (refreshTime && refreshExpression) {
        throw new ConfigurationError(
          `Can not set "refreshTime" and "refreshExpression" at the same time in cache "${cacheTableName}" of schema "${urlPath}"`
        );
      }
      const options = refreshTime
        ? { type: 'refreshTime', timeInterval: refreshTime!.every }
        : { type: 'refreshExpression', timeInterval: refreshExpression!.every };
      const message = this.checkRefreshInterval(options.timeInterval);
      if (message)
        throw new ConfigurationError(
          `The "${options.type}.every" of cache "${cacheTableName}" in schema "${urlPath}" is invalid: ${message}`
        );
    }
  }

  private checkRefreshInterval(timeInterval?: string) {
    if (!timeInterval) return 'Should have a value.';

    const interval = ms(timeInterval as StringValue);
    if (isNaN(interval)) return 'Invalid time string to convert.';
    if (interval < 0) return 'Time can not be negative.';
    return '';
  }
```